More intelligent retry logic on create-team failures

<!--
Clear description of success criteria. Verbosity is ✨!
-->
Despite previously adding some retry logic into `create-team` command, we still randomly get failures on this in our integration tests. And it's not always on the same API calls. It seems there's some kind of race condition in the github backend around creating teams, deleting members, and linking to IdP groups in rapid succession.

The most recent occurrence was this:
```
[2022-08-19T22:26:39.8303455+00:00] [DEBUG] HTTP PATCH: https://api.github.com/orgs/e2e-testing-windows/teams/gei-e2e-2-maintainers/external-groups
[2022-08-19T22:26:39.8519884+00:00] [DEBUG] HTTP BODY: {"group_id":15822}
[2022-08-19T22:26:39.9857320+00:00] [DEBUG] GITHUB REQUEST ID: 0401:3E20:E5E55:1D4A53:63000E1F
[2022-08-19T22:26:39.9859980+00:00] [DEBUG] RESPONSE (BadRequest): {"message":"This team cannot be externally managed since it has explicit members.","documentation_url":"https://docs.github.com/rest/reference/teams#link-external-idp-group-team-connection"}
[2022-08-19T22:26:40.0730493+00:00] [ERROR] System.Net.Http.HttpRequestException: Response status code does not indicate success: 400 (Bad Request).
   at System.Net.Http.HttpResponseMessage.EnsureSuccessStatusCode()
   at OctoshiftCLI.GithubClient.SendAsync(HttpMethod httpMethod, String url, Object body, HttpStatusCode status)
   at OctoshiftCLI.GithubClient.PatchAsync(String url, Object body)
   at OctoshiftCLI.GithubApi.AddEmuGroupToTeam(String org, String teamSlug, Int32 groupId)
   at OctoshiftCLI.Commands.CreateTeamCommandBase.Handle(String githubOrg, String teamName, String idpGroup, String githubPat, Boolean verbose)
```

https://github.com/github/gh-gei/runs/7926726769?check_suite_focus=true#step:18:3323

It seems that even though the delete members code returned success, when we try to map to an IdP group it still thinks the team has members.

We have a few options for how to make this case more resilient:
1. If the IdP linking fails with this error just wait and retry.
2. Try listing and deleting the team members again, then repeat the IdP linking
3. Delete the whole team and start the create-team command over from the start.

Option 1 is probably the best way to go, at least to start.

--- a/RELEASENOTES.md
+++ b/RELEASENOTES.md
@@ -1,1 +1,1 @@
-
+- If `create-team` fails when linking an IdP group with an HTTP 400, we will retry

--- a/src/Octoshift/GithubApi.cs
+++ b/src/Octoshift/GithubApi.cs
@@ -505,7 +505,8 @@ namespace OctoshiftCLI
             var url = $"{_apiUrl}/orgs/{org}/teams/{teamSlug}/external-groups";
             var payload = new { group_id = groupId };
 
-            await _client.PatchAsync(url, payload);
+            await _retryPolicy.HttpRetry(async () => await _client.PatchAsync(url, payload),
+                ex => ex.StatusCode == HttpStatusCode.BadRequest);
         }
 
         public virtual async Task<bool> GrantMigratorRole(string org, string actor, string actorType)

--- a/src/OctoshiftCLI.Tests/GithubApiTests.cs
+++ b/src/OctoshiftCLI.Tests/GithubApiTests.cs
@@ -1233,6 +1233,27 @@ namespace OctoshiftCLI.Tests
         }
 
         [Fact]
+        public async Task AddEmuGroupToTeam_Retries_On_400()
+        {
+            // Arrange
+            const string teamSlug = "TEAM_SLUG";
+            const int groupId = 1;
+
+            var url = $"https://api.github.com/orgs/{GITHUB_ORG}/teams/{teamSlug}/external-groups";
+            var payload = new { group_id = groupId };
+
+            _githubClientMock.SetupSequence(m => m.PatchAsync(url, It.Is<object>(x => x.ToJson() == payload.ToJson()), null))
+                .ThrowsAsync(new HttpRequestException(null, null, HttpStatusCode.BadRequest))
+                .ReturnsAsync("");
+
+            // Act
+            await _githubApi.AddEmuGroupToTeam(GITHUB_ORG, teamSlug, groupId);
+
+            // Assert
+            _githubClientMock.Verify(m => m.PatchAsync(url, It.Is<object>(x => x.ToJson() == payload.ToJson()), null), Times.Exactly(2));
+        }
+
+        [Fact]
         public async Task GrantMigratorRole_Returns_True_On_Success()
         {
             // Arrange

